### DISCUSSION TITLE
permission issue with host devices and ssh

### DISCUSSION THREAD

**Discussion by Viatorus** [containers/podman#20721] [Q&A]
Hello,

I try to access a host device inside a container while being connected through ssh.

I use ```--userns keep-id --group-add keep-groups --device /dev/ttyS0``` when running the container.

Inside the container, I can call `cat /ttyS0` without a problem. `id` shows:
```
uid=1000(newuser) gid=1000(newuser) groups=1000(newuser),65534(nogroup)
```

If I connect me through ssh, I get `Permission denied` wtih `cat /ttyS0`. `id` shows:

```
uid=1000(newuser) gid=1000(newuser) groups=1000(newuser)
```

How do I get the correct group rights when using ssh?

My minimal example:
https://gist.github.com/Viatorus/278266a421443e46bfa53f045a6bd474

**Comment by rhatdan** [comment]
Are you ssh into the container?  Only the primary process (pid 1) of the container is guaranteed access to the leaked group sockets.

**Reply by Viatorus** [reply]
> Are you ssh into the container?
Yes, I ssh into the container.

> Only the primary process (pid 1) of the container is guaranteed access to the leaked group sockets.
Ah, I thought about something like this... :/

But what can I do to make it also work inside a SSH session?

**Reply by Viatorus** [reply]
Are the uid/gid also different? :(

**Reply by rhatdan** [reply]
You could take the hosts GID  and map it to itself within the container. 

If this is a systemd based container running sshd then that would be the only way.

If you are running sshd as PID 1 inside of the container, then you might be able to get sshd to leak the file descriptor to the session, although I doubt you can.

Why are you ssh into the container versus using podman exec?

**Reply by Viatorus** [reply]
> You could take the hosts GID and map it to itself within the container.

Can you explain this in more detail, please?

> Why are you ssh into the container versus using podman exec?

The IDE we are using (CLion/Intellij) works - as I am aware of - only reliable good if the container with the developer environment is connected through ssh. The dev environment but also contains tools like "flash a device over /dev/tty..." which can be directly executed from the IDE.

**Reply by rhatdan** [reply]
Say the GID required for access is GID 10.  Then all users of the tool would need to have 

USER 10:10:1 

In the /etc/subgid.

**Comment by KilinW** [comment]
I encounter the same problem too. We have a NFS mounted NAS that require the user to have specific group to read some data. I'm used to use the container via terminal or dev-container which is totally fine. But some user have IDE that can only with with ssh. And I was unable to make ssh process to have the group that Podman map.
Currently I just simply configure the permission in NAS manually user-by-user without groups. I just wonder if there are any proper method to do this nowadays.

**Comment by eriksjolund** [comment]
I wonder if this is related:

_Note keep-groups can only work if the container itself does not call setgroups()_ 

quote from

https://github.com/containers/podman/discussions/25172#discussioncomment-12016911

Searching for setgroups in openssh code base:

https://github.com/search?q=repo%3Aopenssh%2Fopenssh-portable%20setgroups&type=code

__result:__ setgroups is used in the openssh code base

**Comment by worldxc** [comment]
如果需要sshd服务并且需要用户附加组信息, 可以使用[dropbear](https://github.com/mkj/dropbear)替代openssh. 取消其中的[initgroups](https://github.com/mkj/dropbear/blob/75f699bfe2c234418056776c4d9f651a07a76de6/src/svr-auth.c#L514)调用即可, 这当然有其他安全问题, 如果以rootless方式运行的容器中只有一个root用户也可以接受.
